### ISSUE TITLE
Loading an asset and one of its subassets causes two loads of the same asset to occur.

### ISSUE THREAD

**Issue by andriyDev** [bevyengine/bevy#12756]
## Bevy version

Commit: 6840f95d625b413dcaedb6263eab55dc506525b5

Loading an asset as well as one of its subassets causes the loader to run twice. This also causes asset events to trigger twice, and causes assets to be added to `Assets` and then immediately overwritten.

Here is an example that demonstrates the issue:

```rust
use bevy::{gltf::Gltf, prelude::*};

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Startup, setup)
        .add_systems(Update, (print_mesh_events, print_gltf_events))
        .run();
}

#[derive(Resource)]
struct Handles {
    gltf: Handle<Gltf>,
    mesh: Handle<Mesh>,
}

fn setup(asset_server: Res<AssetServer>, mut commands: Commands) {
    commands.insert_resource(Handles {
        gltf: asset_server.load("scene.gltf"),
        mesh: asset_server.load("scene.gltf#Mesh0/Primitive0"),
    });
}

fn print_mesh_events(mut events: EventReader<AssetEvent<Mesh>>) {
    for event in events.read() {
        dbg!(event);
    }
}

fn print_gltf_events(mut events: EventReader<AssetEvent<Gltf>>) {
    for event in events.read() {
        dbg!(event);
    }
}
```

This prints that any meshes and the gltf are first added and then modified immediately. This likely means the loader is running twice. At the very least, the loader is running once, but the AssetServer loading code is writing to the Assets twice which can be undesirable.

I believe this is because the loaading code in AssetServer sees the asset is not loaded (for that asset path), so it kicks off a load. But since the paths are technically different, it kicks off two loads.

I also tested this with loading two subassets of the same asset, and it also kicks off two loads! Note this doesn't happen if the first load finishes before the second subasset is requested.

**Comment by andriyDev** [comment]
I took a stab at this with #20449 (TL;DR maintain a set of currently loading root asset paths + type IDs, and then bail out of the loader early if there's a load already happening for that path + type ID). That idea is workable, but we come up against a complication: If we are loading an asset (e.g., it takes 10 seconds to load), and a hot reload event occurs, we should stop the existing task and replace it with a new task. I think this means we need to track the notion of "asset version" or something, so that the "old versions" don't prevent the new loader from spinning up (and therefore never loading the asset).

**Comment by andriyDev** [comment]
OK correction to the above: We get a "force" flag, so as long as we adhere to that, it should be ok.

Unfortunately I've run into another issue. There's no way to cancel a load that starts, and the only condition currently is whether an existing load is running. So picture this:

1) We start loading asset#a, and asset#b in the same system.
2) The asset#a task starts, sees there is no task running, loads its asset, and sends it off to the ECS. The system runs, taking the asset and sticking it into `Assets`. This also clears the fact that we're loading an asset.
3) Now asset#b task starts, sees there is no task running, loads its asset, and sends it off to the ECS. blah blah blah same thing happens.

The problem here is that we started both tasks first, and these assume they are loading an asset. Neither gets skipped because the tasks don't run in parallel, they run sequentially. Then they just load the asset as they expect, resulting in two loads.
